etcd3/aio/watch.py

```diff
@@ -0,0 +1,205 @@
+import asyncio
+import logging
+
+import grpc
+
+import etcd3.etcdrpc as etcdrpc
+import etcd3.events as events
+import etcd3.exceptions as exceptions
+import etcd3.utils as utils
+
+
+_log = logging.getLogger(__name__)
+
+
+def create_task(coro):
+    asyncio.get_event_loop().create_task(coro)
+
+
+class Watch(object):
+
+    def __init__(self, watch_id, iterator=None, etcd_client=None):
+        self.watch_id = watch_id
+        self.etcd_client = etcd_client
+        self.iterator = iterator
+
+    async def cancel(self):
+        await self.etcd_client.cancel_watch(self.watch_id)
+
+    def iterator(self):
+        if self.iterator is not None:
+            return self.iterator
+
+        raise ValueError('Undefined iterator')
+
+
+class Watcher(object):
+
+    def __init__(self, watchstub, timeout=None, call_credentials=None,
+                 metadata=None):
+        self.timeout = timeout
+        self._watch_stub = watchstub
+        self._credentials = call_credentials
+        self._metadata = metadata
+
+        self._lock = asyncio.Lock()
+        self._request_queue = asyncio.Queue(maxsize=10)
+        self._callbacks = {}
+        self._callback_thread = None
+        self._new_watch_cond = asyncio.Condition(lock=self._lock)
+        self._new_watch = None
+
+    async def add_callback(self, key, callback, range_end=None,  # noqa: C901
+                           start_revision=None, progress_notify=False,
+                           filters=None, prev_kv=False):
+        create_watch = etcdrpc.WatchCreateRequest()
+        create_watch.key = utils.to_bytes(key)
+        if range_end is not None:
+            create_watch.range_end = utils.to_bytes(range_end)
+        if start_revision is not None:
+            create_watch.start_revision = start_revision
+        if progress_notify:
+            create_watch.progress_notify = progress_notify
+        if filters is not None:
+            create_watch.filters = filters
+        if prev_kv:
+            create_watch.prev_kv = prev_kv
+        rq = etcdrpc.WatchRequest(create_request=create_watch)
+
+        async with self._lock:
+            # Start the callback thread if it is not yet running.
+            if not self._callback_thread:
+                self._callback_thread = create_task(self._run())
```

Comment:
it would be good to apply a little bit of structured concurrency here

```python
async def __aenter__(self):
    self._current_task = asyncio.current_task()
    self._callback_thread = create_task(self._run())

async def __aexit__(self):
    assert self._current_task is asyncio.current_task()

    self._current_task.cancel()
    try:
        await self._callback_thread
    except asyncio.CancelledError:
        pass
```

Reply:
```python
class Watcher:
    def __init__(self, request_queue, watchstub, timeout=None, call_credentials=None,
                 metadata=None):
        self.timeout = timeout
        self._watch_stub = watchstub
        self._credentials = call_credentials
        self._metadata = metadata

        self._lock = anyio.Lock()
        self._request_queue = request_queue
        self._callbacks = {}
        self._callback_thread = None
        self._new_watch_cond = anyio.Condition(lock=self._lock)
        self._new_watch = None
        self._callback_started = False

    def add_callback(self):
        ...
       
             if not self._callback_started:
                 self._callback_started = True
                 self._task_group.start_soon(self._run())

    @contextlib.asynccontextmanager
    async def create(cls, watchstub, timeout=None, call_credentials=None,
                 metadata=None):
        async with anyio.create_task_group() as tg, anyio.create_memory_object_stream(10) as rq:
            v = cls(tg, rq, watchstub, timeout, call_credential)